File: lib/scrapers/salcobrand.ts
import type { ScrapedProduct } from "./types";

const APP_ID  = "GM3RP06HJG";
const API_KEY = "0259fe250b3be4b1326eb85e47aa7d81";
const INDEX   = "sb_variant_production";
const BASE    = "https://salcobrand.cl";

export async function searchSalcobrand(query: string): Promise<ScrapedProduct[]> {
  try {
    const res = await fetch(
      `https://${APP_ID}-dsn.algolia.net/1/indexes/${INDEX}/query`,
      {
        method: "POST",
        headers: {
          "X-Algolia-Application-Id": APP_ID,
          "X-Algolia-API-Key": API_KEY,
          "Referer": `${BASE}/`,
          "Origin": BASE,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ query, hitsPerPage: 10 }),
      }
    );
    const data = await res.json();
    return (data.hits ?? []).flatMap((hit: Record<string, unknown>) => {
      const normal   = hit.normal_price      as number | null;
      const internet = hit.internet_price    as number | null;  // precio web
      const cmr      = hit.cmr_price         as number | null;  // precio tarjeta CMR
      const direct   = hit.direct_discount   as string | null;  // fallback descuento

      const storePrice = normal ? Number(normal) : null;
      if (!storePrice) return [];

      // Algolia solo expone normal_price y direct_discount.
      // CMR e internet solo están en la página de producto (requieren sesión).
      const directNum = direct ? parseFloat(direct) : null;
      const onlinePrice = directNum && directNum < storePrice ? directNum : null;

      // internet_price y cmr_price no están disponibles en el índice de búsqueda
      const cmrPrice = cmr ? Number(cmr) : null;

      const name = String(hit.name ?? query);
      const slug = (hit.slug as string) ?? "";
      const sku  = (hit.sku  as string) ?? "";
      const bio  = (hit.bioequivalent_filter as Record<string, unknown>) ?? {};
      const url  = slug && sku ? `${BASE}/products/${slug}?default_sku=${sku}` : (slug ? `${BASE}/products/${slug}` : null);

      return [{
        name,
        price: storePrice,
        onlinePrice,
        cmrPrice,
        hasStock: Boolean(hit.has_stock ?? true),
        hasOnlineDelivery: Boolean(hit.package_delivery ?? true),
        onlineUrl: url,
        laboratory: (hit.brand as string) ?? null,
        isBioequivalent: Boolean(bio.has_bioequivalent ?? false),
      }];
    });
  } catch {
    return [];
  }
}

File: scraper/scrapers/salcobrand.py
import httpx
from .base import BaseScraper, ScrapedProduct

ALGOLIA_APP_ID  = "GM3RP06HJG"
ALGOLIA_API_KEY = "0259fe250b3be4b1326eb85e47aa7d81"
ALGOLIA_INDEX   = "sb_variant_production"
BASE_URL        = "https://salcobrand.cl"


class SalcobrandScraper(BaseScraper):
    slug = "salcobrand"
    name = "Salcobrand"

    async def search(self, query: str) -> list[ScrapedProduct]:
        results = []
        try:
            async with httpx.AsyncClient(timeout=15) as client:
                r = await client.post(
                    f"https://{ALGOLIA_APP_ID}-dsn.algolia.net/1/indexes/{ALGOLIA_INDEX}/query",
                    json={"query": query, "hitsPerPage": 10},
                    headers={
                        "X-Algolia-Application-Id": ALGOLIA_APP_ID,
                        "X-Algolia-API-Key": ALGOLIA_API_KEY,
                        "Referer": f"{BASE_URL}/",
                        "Origin": BASE_URL,
                        "User-Agent": "Mozilla/5.0",
                    },
                )
                r.raise_for_status()
                hits = r.json().get("hits", [])

            for hit in hits:
                # normal_price = precio internet (único disponible vía Algolia)
                # direct_discount = descuento adicional (si existe y es menor)
                normal = hit.get("normal_price")
                price = float(normal) if normal else None
                if not price:
                    continue

                direct = hit.get("direct_discount")
                direct_num = float(direct) if direct else None
                # online_price solo si hay un descuento menor al precio base
                online_price = direct_num if direct_num and direct_num < price else None

                slug = hit.get("slug", "")
                sku  = hit.get("sku", "")
                bio  = hit.get("bioequivalent_filter", {})
                url  = f"{BASE_URL}/products/{slug}?default_sku={sku}" if slug and sku else (f"{BASE_URL}/products/{slug}" if slug else None)

                results.append(ScrapedProduct(
                    name=hit.get("name", query),
                    price=price,
                    online_price=online_price,
                    has_stock=bool(hit.get("has_stock", True)),
                    has_online_delivery=bool(hit.get("package_delivery", True)),
                    online_url=url,
                    laboratory=hit.get("brand"),
                    is_bioequivalent=bool(bio.get("has_bioequivalent", False)),
                ))
        except Exception as e:
            print(f"[Salcobrand] Error: {e}")
        return results
